File: src/CollectIt.API/CollectIt.API.Tests.Integration/AuthorizationControllerTests.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json;
using System.Threading.Tasks;
using CollectIt.API.DTO;
using CollectIt.Database.Infrastructure;
using Xunit;
using Xunit.Abstractions;

namespace CollectIt.API.Tests.Integration;

[Collection("Open id dict tests")]
public class AuthorizationControllerTests : IClassFixture<CollectItWebApplicationFactory>
{
    private readonly CollectItWebApplicationFactory _factory;
    private readonly ITestOutputHelper _testOutputHelper;

    public AuthorizationControllerTests(CollectItWebApplicationFactory factory, ITestOutputHelper testOutputHelper)
    {
        _factory = factory;
        _testOutputHelper = testOutputHelper;
    }

    [Fact]
    public async Task GetToken_WithValidUsernameAndPassword_ShouldReturnIdToken()
    {
        var admin = PostgresqlCollectItDbContext.AdminUser;
        using var client = _factory.CreateClient();
        using var message = new HttpRequestMessage(HttpMethod.Post, "/connect/token");
        message.Content = new FormUrlEncodedContent(new[]
                                                    {
                                                        new KeyValuePair<string, string>("grant_type", "password"),
                                                        new KeyValuePair<string, string>("username", admin.UserName),
                                                        new KeyValuePair<string, string>("password", "12345678")
                                                    });
        
        var response = await client.SendAsync(message);
        var result = await response.Content.ReadFromJsonAsync<AccountDTO.OpenIddictResponseSuccess>();
        Assert.NotNull(result);
        Assert.Equal("Bearer", result.TokenType);
        Assert.NotEmpty(result.AccessToken);
    }
    
    private async Task<(HttpClient, string)> Initialize(string? username = null, string? password = null)
    {
        var client = _factory.CreateClient();
        var bearer = await TestsHelpers.GetBearerForUserAsync(client, 
                                                              username: username, 
                                                              password: password);
        return ( client, bearer );
    }

    [Fact]
    public async Task Register_WithValidValidValues_ShouldRegisterUser()
    {
        using var client = _factory.CreateClient();
        var username = "SomeValidUsername";
        var password = "SomeP@ssw0rd";
        var email = "test@mail.ru";
        var message = new HttpRequestMessage(HttpMethod.Post, "connect/register")
                      {
                          Content = new FormUrlEncodedContent(new KeyValuePair<string, string>[]
                                                              {
                                                                  new ("email", email),
                                                                  new ("password", password),
                                                                  new ("username", username)
                                                              })
                      };
        var response = await client.SendAsync(message);
        Assert.Equal(HttpStatusCode.Created, response.StatusCode);
        var actual = await  response.Content.ReadFromJsonAsync<AccountDTO.ReadUserDTO>();
        Assert.NotNull(actual);
        Assert.Equal(username, actual!.UserName);
        Assert.Equal(email, actual.Email);
    }
    
    [Fact]
    public async Task CreateUser_AlreadyRegisteredUsername_ShouldReturnBadRequest()
    {
        using var client = _factory.CreateClient();
        var user = PostgresqlCollectItDbContext.AdminUser;
        var dto = new AccountDTO.CreateUserDTO(user.UserName, "oleg@mail.ru", "P@ssw0rd2H@rd");
        using var message = new HttpRequestMessage(HttpMethod.Post, "connect/register")
                            {
                                Content = new FormUrlEncodedContent(new KeyValuePair<string, string>[]
                                                                    {
                                                                        new("username", dto.UserName),
                                                                        new("email", dto.Email),
                                                                        new("password", dto.Password)
                                                                    }),
                            };
        var response = await client.SendAsync(message);
        Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
    }
}

File: src/CollectIt.API/CollectIt.API.Tests.Integration.FSharp/AuthorizationControllerTests.fs
module Tests

open System.Collections.Generic
open System.Net
open System.Net.Http
open System.Net.Http.Json
open System.Text.Json
open CollectIt.API.DTO
open CollectIt.API.DTO.AccountDTO
open CollectIt.API.Tests.Integration.FSharp.CollectItWebApplicationFactory
open CollectIt.Database.Infrastructure
open Xunit
open Xunit.Abstractions

[<Collection("OpenIddict tests")>]
type AuthorizationControllerTests(factory: CollectItWebApplicationFactory, output: ITestOutputHelper) =
    class
        member this._factory = factory
        member this._output = output
        member private this.log msg = this._output.WriteLine msg

        interface IClassFixture<CollectItWebApplicationFactory>

        [<Fact>]
        member this.``Should get valid token with admin username and password``() =
            async {
                let admin =
                    PostgresqlCollectItDbContext.AdminUser

                use client = this._factory.CreateClient()

                let content =
                    new FormUrlEncodedContent(
                        [ KeyValuePair("grant_type", "password")
                          KeyValuePair("username", admin.UserName)
                          KeyValuePair("password", "12345678") ]
                    )

                use message =
                    new HttpRequestMessage(HttpMethod.Post, "/connect/token", Content = content)

                let log x = this.log (x.ToString())
                log 1

                let! response =
                    client.SendAsync(message, Async.DefaultCancellationToken)
                    |> Async.AwaitTask

                log 2

                let! result =
                    HttpContentJsonExtensions.ReadFromJsonAsync<OpenIddictResponseSuccess>(
                        response.Content,
                        JsonSerializerOptions(JsonSerializerDefaults.Web),
                        Async.DefaultCancellationToken
                    )
                    |> Async.AwaitTask

                log 3
                Assert.NotNull result
                Assert.Equal("Bearer", result.TokenType)
                Assert.NotEmpty result.AccessToken
            }
        //            |> Async.AwaitTask
//            |> Async.RunSynchronously

        [<Fact>]
        member this.``Register user with valid username, email and password should create new user``() =
            use client = this._factory.CreateClient()
            let username = "SomeValidUsername"
            let password = "SomeP@ssw0rd"
            let email = "test@mail.ru"

            let form =
                new FormUrlEncodedContent(
                    [ KeyValuePair("username", username)
                      KeyValuePair("password", password)
                      KeyValuePair("email", email) ]
                )

            let message =
                new HttpRequestMessage(HttpMethod.Post, "/connect/register", Content = form)

            task {

                let! response = client.SendAsync message
                Assert.Equal(HttpStatusCode.Created, response.StatusCode)
                let! actual = HttpContentJsonExtensions.ReadFromJsonAsync<AccountDTO.ReadUserDTO> response.Content
                Assert.NotNull actual
                Assert.Equal(username, actual.UserName)
                Assert.Equal(email, actual.Email)
            }



    end
